Firestore/core/test/firebase/firestore/util/async_queue_test.cc

```diff
@@ -0,0 +1,246 @@
+  EXPECT_ANY_THROW(Queue().VerifyIsCurrentQueue());
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueRequiresOperationInProgress) {
+  auto queue = Queue();
+  dispatch_sync_f(underlying_queue, &queue, [](void* const raw_queue) {
+    EXPECT_ANY_THROW(
+        static_cast<AsyncQueue*>(raw_queue)->VerifyIsCurrentQueue());
+  });
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueWorksWithOperationInProgress) {
+  auto queue = Queue();
+  queue.RunSync([&] { EXPECT_NO_THROW(queue.VerifyIsCurrentQueue()); });
+}
+
+TEST(AsyncQueue, CanScheduleOperationsInTheFuture) {
+  auto signal_finished = CreateSignal();
+  std::string steps;
+
+  auto queue = Queue();
+  queue.Enqueue([&steps] { steps += '1'; });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), kTimerId1,
+                          [&steps, &signal_finished] {
+                            steps += '4';
+                            signal_finished();
+                          });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(1), kTimerId2,
+                          [&steps] { steps += '3'; });
+  queue.Enqueue([&steps] { steps += '2'; });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+  EXPECT_EQ(steps, "1234");
+}
+
+TEST(AsyncQueue, CanCancelDelayedCallbacks) {
+  auto signal_finished = CreateSignal();
+  std::string steps;
+
+  auto queue = Queue();
+  queue.Enqueue([&] {
+    // Queue everything from the queue to ensure nothing completes before we
+    // cancel.
+
+    queue.EnqueueAllowingSameQueue([&steps] { steps += '1'; });
+
+    DelayedOperation delayed_operation = queue.EnqueueAfterDelay(
+        AsyncQueue::Milliseconds(1), kTimerId1, [&steps] { steps += '2'; });
+
+    queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), kTimerId2,
+                            [&steps, &signal_finished] {
+                              steps += '3';
+                              signal_finished();
+                            });
+
+    EXPECT_TRUE(queue.ContainsDelayedOperation(kTimerId1));
+    delayed_operation.Cancel();
+    // Note: the operation will only be removed from the queue after it's run,
+    // not immediately once it's canceled.
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+  EXPECT_EQ(steps, "13");
+}
+
+TEST(AsyncQueue, DelayedOperationIsValidAfterTheOperationHasRun) {
+  auto signal_finished = CreateSignal();
+
+  auto queue = Queue();
+  DelayedOperation delayed_operation = queue.EnqueueAfterDelay(
+      AsyncQueue::Milliseconds(1), kTimerId1, [&] { signal_finished(); });
+  EXPECT_TRUE(queue.ContainsDelayedOperation(kTimerId1));
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+  EXPECT_FALSE(queue.ContainsDelayedOperation(kTimerId1));
+  EXPECT_NO_THROW(delayed_operation.Cancel());
+}
+
+TEST(AsyncQueue, CanManuallyDrainAllDelayedCallbacksForTesting) {
+  std::string steps;
+
+  auto queue = Queue();
+  queue.Enqueue([&steps] { steps += '1'; });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(20000), kTimerId1,
```

Comment:
technically, as above. But if we're context switching for 10s+, we've got bigger problems.